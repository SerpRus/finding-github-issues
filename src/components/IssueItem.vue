<template>
  <li
    class="issue-item"
  >
    <div
      class="issue-item__state"
    >
      State:
      <span
        class="issue-item__state-value"
        v-bind:class="[{'issue-item__state-value--closed': isStateClosed}]"
      >
        {{ value.state }}
      </span>
    </div>

    <a
      class="issue-item__link"
      :href="value.html_url"
    >
      {{ value.title }}
    </a>

    <div class="issue-item__info">
      <div class="issue-item__number-with-date">
        <div class="issue-item__number">
          #{{ value.number }}
        </div>

        <div class="issue-item__date">
          opened on {{ getDate() }}
        </div>
      </div>

      <div class="issue-item__author">
        by

        <a
          class="issue-item__login"
          :href="value.user.html_url"
        >
          {{ value.user.login }}
        </a>
      </div>
    </div>
  </li>
</template>

<script>

export default {
  name: 'IssueItem',
  props: {
    value: Object,
  },
  data() {
    return {
      isStateClosed: this.value.state === 'closed',
    };
  },
  methods: {
    getDate() {
      const date = new Date(this.value.created_at);
      const formatter = new Intl.DateTimeFormat('en', { month: 'short' });

      const dayOfMonth = date.getDate();
      const month = formatter.format(date);

      return `${dayOfMonth} ${month}`;
    },
  },
};
</script>

<style lang="scss" scoped>
.issue-item {
  margin: 0;
  padding: 5px 10px;

  list-style: none;

  border-top: 2px solid brown;
  border-right: 2px solid brown;
  border-left: 2px solid brown;

  @media (min-width: 767px) {
    padding: 10px 15px;
  }

  &:first-child {
    border-radius: 8px 8px 0 0;
  }

  &:last-child {
    border-bottom: 2px solid brown;
    border-radius: 0 0 8px 8px;
  }

  &__state {
    padding: 5px 0;

    &-value {
      color: green;

      &--closed {
        color: red;
      }
    }
  }

  &__link {
    display: inline-flex;

    margin: 5px 0;

    color: #000;

    &:hover {
      color: brown;
    }

    &:focus {
      outline: 2px solid orange;
    }
  }

  &__info {
    margin: 5px 0;

    @media (min-width: 767px) {
      display: flex;
    }
  }

  &__number-with-date,
  &__author {
    display: flex;
  }

  &__number-with-date {
    @media (min-width: 767px) {
      padding-right: 5px;
    }
  }

  &__number {
    padding-right: 5px;
  }

  &__login {
    margin-left: 5px;

    color: #000;

    &:hover {
      color: brown;
    }

    &:focus {
      outline: 2px solid orange;
    }
  }
}
</style>
