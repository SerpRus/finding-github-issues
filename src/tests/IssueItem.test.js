import {mount} from '@vue/test-utils';
import IssueItem from '../components/IssueItem.vue';

describe('IssueItem tests', () => {
  const data = {
    html_url: 'https://github.com/vuejs/vue/pull/12737',
    number: 12737,
    state: 'open',
    created_at: '2022-08-11T09:27:06Z',
    title: 'feat(types): new Vue() improvements (#12730)',
    user: {
      login: 'gulewei',
      html_url: 'https://github.com/gulewei'
    },
  };

  const wrapper = mount(IssueItem, {
    props: {
      value: data,
    },
  });

  it('check state', () => {
    expect(wrapper.find('.issue-item__state-value').text()).toBe(data.state);
  });

  it('check link', () => {
    expect(wrapper.find('.issue-item__link').attributes().href).toBe(data.html_url);
    expect(wrapper.find('.issue-item__link').text()).toBe(data.title);
  });

  it('check number', () => {
    expect(wrapper.find('.issue-item__number').text()).toBe(`#${data.number}`);
  });

  it('check date', () => {
    expect(wrapper.find('.issue-item__date').text()).toBe('opened on 11 Aug');
  });

  it('check author', () => {
    expect(wrapper.find('.issue-item__login').attributes().href).toBe(data.user.html_url);
    expect(wrapper.find('.issue-item__login').text()).toBe(data.user.login);
  });

  it('IssueItem snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
