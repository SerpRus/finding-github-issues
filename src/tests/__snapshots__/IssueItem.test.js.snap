// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueItem tests IssueItem snapshot 1`] = `
"<li class=\\"issue-item\\">
  <div class=\\"issue-item__state\\"> State: <span class=\\"issue-item__state-value\\">open</span></div><a class=\\"issue-item__link\\" href=\\"https://github.com/vuejs/vue/pull/12737\\">feat(types): new Vue() improvements (#12730)</a>
  <div class=\\"issue-item__info\\">
    <div class=\\"issue-item__number-with-date\\">
      <div class=\\"issue-item__number\\"> #12737</div>
      <div class=\\"issue-item__date\\"> opened on 11 Aug</div>
    </div>
    <div class=\\"issue-item__author\\"> by <a class=\\"issue-item__login\\" href=\\"https://github.com/gulewei\\">gulewei</a></div>
  </div>
</li>"
`;
